### ISSUE TITLE
Critical Bug - iOS - PurchaseAsync also Auto-Consumes the Purchase.

### ISSUE THREAD

**Issue by najak3d** [jamesmontemagno/InAppBillingPlugin#412]
## Bug Information

Version Number of Plugin: 4.0.2
Device Tested On:   iPad Mini 3,  iOS 12.4
Simulator Tested On: n/a
Version of VS: VS for Mac 8.10.8
Version of Xamarin:   6.18.0.23
Versions of other things you are using:

### Steps to reproduce the Behavior
Call   "await PurchaseAsync(...)" and complete the purchase transaction in the AppStore UI (i.e. enter password, and tap "Buy").

### Expected Behavior
It should ONLY initiate the Purchase and return the Purchase Item with Receipt, BUT should NOT auto-CONSUME the item from the AppStore.   That purchase item should remain in the AppStore list until we call "ConsumePurchaseAsync".

### Actual Behavior
After you purchase the item, this InAppBilling plugin auto-CONSUMES this item, so that it no longer appears in the store.
Therefore, if you call "ConsumePurchaseAsync" on this plugin, it throws a null error (as there is nothing to consume).

To confirm this behavior, I have also attached an "SKPaymentTransactionObserver" which listens to all of the events.  And sure enough, immediately following a successful AppStore purchase, we are notified of the purchase, and THEN it immediately calls "RemovedTransactions" on this observer, showing the transaction has been removed from the AppStore.   Future queries of the AppStore show it missing... it was "consumed/removed".

For our app's logic to function here (the same as we have it working for Android), we are counting on that item to remain on the AppStore until after we have Installed the Consumable in our Application for the user.   Once we're done, we then call ConsumePurchaseAsync to remove it from the store.

Other side-effects of this appear as:
1. QueryPurchases() - always returns empty list!
2. AcknowledgePurchase - always returns "true" (even for an invalid PurchaseToken)
3. ConsumePurchaseAsync() - always throws a null error.

So the only way we can make this plugin work on iOS, is to ONLY CALL PurchaseAsync() - and then handle the Happy Path logic on the callback to "OnPurchaseComplete()" event handler.    This is non-sufficient, as it does not enable us to handle the non-happy path failure modes at risk.   Plus we'd like to have a shared store logic between Android and iOS, instead of something "very different".

### Code snippet
await _serviceConnection.PurchaseAsync(productId, ItemType.Consumable, this);   //  "this" implements the "VerifyPurchase"

**Comment by jamesmontemagno** [comment]
Checkout the docs -> https://jamesmontemagno.github.io/InAppBillingPlugin/PurchaseConsumable.html

On iOS this is how it works. You only need to call "ConsumePurchase" on Android/Windows. This is just how the iOS App store works.

**Comment by najak3d** [comment]
For iOS, we created a mock-plugin using the SKPaymentObserver - so that it works the SAME as it does on Android.   It keeps the purchases on the Store UNTIL we call Consume.   This way we have a safety fallback, in case we fail to handle the "PurchaseComplete" event.

I believe the InAppBilling plugin should follow this same model.  Do not call "finishTransaction" until the app calls ConsumePurchase() on your plugin.

The flow we have for iOS is:
1. PurchaseAsync() -- pays for it, and puts it into the AppStore list.    <== Our App can safely crash right here.
2. QueryPurchases() - shows this purchased item, UNTIL we consume it.
3. We then Apply the purchase to the App, and on our server - so that they user gets their benefit.
4. THEN, we call Consume, and it removes it from the Store.

===
While, if we used your Plugin for iOS - we'd be stuck with this:
1. PurchaseAsync() ->  then triggers the "PurchaseComplete" event AND Consumes it.  <== App Crash, means we lose the purchase!
2. IF we mistakenly lose this event, and are unable to Apply/Install the purchase ... we have less of a fallback.

If our App crashes, or the install process fails first time -- we have no way to query the store for unconsumed purchases.

===
Thus we've created a mock plugin (same API as yours) for iOS - which "works equivalent to Android" (except for Acknowledge, which we just "pretend it works", since this one concept is not part of AppStore).

I can send you our plugin code (1 file) that accomplishes this, if you'd like to see it.

**Comment by najak3d** [comment]
I checked out your code here:
https://github.com/jamesmontemagno/app-ac-islandtracker/blob/master/TurnipTracker/ViewModel/ProViewModel.cs

Line 132 - you call PurchaseAsync().

Which for a Consumable item, will ALSO auto-consume it.   Now it's gone from the store, and totally in the hands of your application.

Now imagine that something goes wrong in your App from lines 134 to 166 where you call:
DataService.CreateProStatus(...)

So if your app drops-the-ball at all, the consumable is now fully GONE.  The user will not get what they paid for.

ALSO, we run the risk of a "Pending Payment" where it comes back as Non-Null, but is "Pending".   Now imagine that 2 seconds later, the purchase Completes (becomes "Purchased"), but your App will NEVER see it.    So here, the user pays for it, but NEVER gets what they paid for.   How do you deal with the "Pending Payment" scenario?   (Status:  Deferred?)

====
In our App, in my Mock  Plugin, made to work "almost the same as yours does for Android", we do NOT auto-Consume the item.   We leave it safely in the TransactionQueue, where it sits until we decide to call "FinishTransaction" (via our mock API of ConsumeAsync()).     We only call this AFTER we've ensured a full installation of the product (and our server has validated the Receipt with the AppStore).

In our App, if the purchase stays "Deferred" (or pending) -- we'll still pick it up on the next restart of the App... as we call "QueryPurchases()" and the item will show up, and then we continue our process of installation, and consumption.

I'm sure your code works well for the "Happy Path" (which means it doesn't return from PurchaseAsync() until it's either "purchased" or "cancelled/failed", AND your App handles it without issue)...     For us, the Happy Path on iOS only seems to happen about 95% of the time, which means we'd end up with 1 in 20 not getting what they paid for.   Our consumables cost $80 to $150 each -- so you can imagine the customer disappointment when it fails.

Thus we are forced to use more robust logic, to correctly handle the 5% failure rate.  If you want to see our code, I'd be happy to send it to you.

**Comment by jamesmontemagno** [comment]
Gotcha, I see, so we need a flag for iOS to know if it is a consumable or not and if they want to auto-consume or not. I'll take a look at this.

**Comment by najak3d** [comment]
> Gotcha, I see, so we need a flag for iOS to know if it is a consumable or not and if they want to auto-consume or not. I'll take a look at this.

That would be awesome.  Thank you for your hard and good work here.

The trickiest thing we had to do in our mock-plugin for iOS was to deal with a different paradigm in how the StoreKit deals with the "Purchase()" request -- it does NOT return a purchase.  It only launches it..   The "TransactionsUpdated()" event will later get fired on the SKObserver, letting us know the status of this item...  

1. It fires immediately showing the purchase item with status "Purchasing".
2. Then from there it'll get updated again whenever status changes.
3. For Failure -- we Auto-Consume this.
4. When it transitions to "Purchased" that's when our Mock-Plugin "await Purchase()" call will then return to us with the Transaction object (which we translate to InAppBillingPurchase equivalent.  So that it appears to our App the same as your plugin.

Since UpdatedTransactions() - sends a list -- if you have pre-existing purchases in the queue that haven't been completed (maybe they are pending, or you just aren't ready to consume them yet) - we set a Field call "AwaitingPurchaseOf" string set to the ProductId of the purchase we are expecting to show up as "Purchasing" state.   This is how we identity which specific TransactionID that is to be sent back to the "PurchaseAsyn()" method.

That was the tricky part - to make sure this worked reliably in the case of having pre-existing non-consumed purchases in the payment queue.

Our Mock-plugin for iOS is now fully tested, and about 500 lines of code, and handles what I believe are the most adverse situations/complexities.  If you want to see our code, I can send it to you for reference.
